### PR TITLE
Ports - Margaret 

### PR DESCRIPTION
**Pull request by msfinnan**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | The spotlight model method finds the work with the most votes
Describe how you approached testing that model method. What edge cases did you come up with? | I used the fixtures to test nominal cases where there are works with votes. I deleted the fixtures and created my own instances of works to test an edge case where there are no works with votes (to ensure it still show something)  and to test that it will still show one title when there is a tie in the number of votes. 
What are `session` and `flash`? What is the difference between them? | Session is a way to keep track of a user using the app, flash is way to display messages to a user. Flash is persistent through current or 1request/response cycles. Session uses cookies and is persistent through many request/response cycles. 
What was one thing that you gained more clarity on through this assignment? | Custom model methods, testing models, and fixtures.  I was not very confident on these going into the project. This project helped me get a little more comfortable with custom model methods and coming up with ways to test them, when to use fixtures, and when to create my own test subjects. 
What is the Heroku URL of your deployed application? | https://margis-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | I don't know of a way to teach this, but I'm still very uncomfortable troubleshooting Heroku, and think it would be great to have a little more knowledge about that going into the project.

### CODE REVIEW

File: test/controllers/users_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  describe "current" do
+    it "returns 200 OK for a logged-in user" do
+      # Arrange
+      perform_login
+
+      # Act
+      get current_user_path
+
+      # Assert
+      must_respond_with :success
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb]: When the user logs in it redirects them, not responding with success.

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    end
+
+    it "respond with redirect if user is not logged in" do
+      # Arrange
+      # no user logged in / could log in and then log out post logout_path
+      # Act
+      get current_user_path
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb]: This path doesn't exist, instead maybe the user show path?

File: test/controllers/votes_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      vote_hash = {
+        vote: {
+          user_id: user.id,
+          work_id: work.id,
+        },
+      }
+
+      # Act-Assert
+      expect {
+        post work_votes_path(work.id), params: vote_hash
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:19]: If you notice there's no form in the upvote action.  Instead you just make a post request and the controller figures out the user by session and work by the route parameter.  

Also remember to test your flash notices.

File: test/controllers/votes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      }.must_change "Vote.count", 1
+
+      # try to vote on same work again
+      expect {
+        post work_votes_path(work.id), params: vote_hash
+      }.wont_change "Vote.count"
+
+      must_respond_with :redirect
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:48]: You should also test flash notices.

File: test/controllers/works_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        post works_path, params: work_hash
+      }.wont_change "Work.count"
+
+      # Assert
+      must_respond_with :bad_request
+    end
+  end
+
+  describe "edit" do
+    it "can get the edit work page" do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:103]: What about the edit action without a valid `work_id`

File: test/controllers/works_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+      must_respond_with :redirect
+      work_to_update.reload
+      expect(work_to_update.title).must_equal work_hash[:work][:title]
+      expect(work_to_update.description).must_equal work_hash[:work][:description]
+      expect(work_to_update.publication_year).must_equal work_hash[:work][:publication_year]
+      expect(work_to_update.creator).must_equal work_hash[:work][:creator]
+      expect(work_to_update.category).must_equal work_hash[:work][:category]
+    end
+
+    it "will return a 400 when asked to update with invalid data" do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:143]: You should also have a test with an invalid work_id

File: test/models/work_test.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  it "must be valid" do
+    work = works(:work_one)
+
+    valid_work = work.valid?
+
+    expect(valid_work).must_equal true
+  end
+
+  describe "validations" do
+    it "requires a title" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:15]: It would be better if you also had uniqueness as a validation for a title.

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+      # Assert
+      expect(valid_work).must_equal false
+      expect(work.errors.messages).must_include :category
+      expect(work.errors.messages[:category]).must_equal ["is not included in the list"]
+    end
+  end
+
+  describe "sort_work model method" do
+    it "sorts work of a certain category by votes" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:56]: You should also test this with other categories and invalid categories

File: test/models/work_test.rb [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+      # Act-Assert
+      expect(Work.highest_vote).wont_be_nil
+    end
+
+    it "will spotlight one work if there is a tie for most votes" do
+      # Arrange
+      Vote.delete_all
+
+      vote = Vote.new(user_id: users(:user_one), work_id: works(:work_one))
```
Comment (CheezItMan) [inline test/models/work_test.rb]: You made votes here, but you didn't **`save`** them!

File: test/models/work_test.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+    it "spotlights a work even if no works have votes" do
+      # Arrange
+      Vote.delete_all
+
+      # Act-Assert
+      expect(Work.highest_vote).wont_be_nil
+    end
+
+    it "will spotlight one work if there is a tie for most votes" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:76]: What happens if there are no works?

File: test/models/work_test.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    end
+  end
+
+  describe "top_ten model method" do
+    it "shows ten works of a certain type with the most votes" do
+      # Act-Assert
+      expect(Work.top_ten("album").first).must_equal works(:work_ten)
+    end
+
+    it "shows 10 works even when there are less than 10 works with votes" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:96]: You should have this method when there are less than 10 works, and more than 10.

File: test/models/vote_test.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+require "test_helper"
+
+describe Vote do
+  it "must be valid" do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:4]: You should also have a validation to enforce a user & work combination has to be unique.

File: app/models/work.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
+class Work < ApplicationRecord
+  has_many :votes, dependent: :destroy
+  validates :title, presence: true
+  validates :category, presence: true, inclusion: {in: %w(album book movie)}
+
+  def self.sort_work(category)
+    works = Work.where(category: category).to_a
+    works.sort_by! { |work| Vote.where(work_id: work.id).length }
```
Comment (CheezItMan) [inline app/models/work.rb:8]: Just a note, it's more efficient to sort things using the database, using `order`

File: app/models/work.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
+class Work < ApplicationRecord
```
Comment (CheezItMan) [inline app/models/work.rb:1]: I also suggest adding an `upvote` method to work, which takes a user and tries to add a vote to the current work.

File: app/models/vote.rb [hunk 15]
```diff
@@ -0,0 +1,4 @@
+class Vote < ApplicationRecord
+  belongs_to :user
```
Comment (CheezItMan) [inline app/models/vote.rb:2]: validations?

File: app/controllers/works_controller.rb [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    # @work = Work.find_by(id: params[:id])
+    if @work.nil?
+      redirect_to works_path
+    end
+  end
+
+  def update
+    # work = Work.find_by(id: params[:id])
+
+    updated_successfully = @work.update(work_params)
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:48]: What about if you can't find the work?

File: app/controllers/votes_controller.rb [hunk 17]
```diff
@@ -0,0 +1,22 @@
+class VotesController < ApplicationController
+  def create
+    @vote = Vote.new(user_id: session[:user_id], work_id: params[:work_id])
+
+    if Vote.find_by(user_id: session[:user_id], work_id: params[:work_id]) != nil
```
Comment (CheezItMan) [inline app/controllers/votes_controller.rb:5]: Notice how complicated this method is.  Putting the business logic into a model method would be appropriate.

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def login
+    username = params[:user][:username]
+    user = User.find_by(username: username)
+    if user #if is valid
+      session[:user_id] = user.id
+      flash[:success] = "Successfully logged in as returning user #{username}"
+    else
+      user = User.create(username: username)
```
Comment (CheezItMan) [inline app/controllers/users_controller.rb:13]: What if `create` fails?

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+    if @user.nil?
+      flash[:error] = "Unknown user"
+
+      redirect_to users_path
+    end
+  end
+
+  def logout
+    session[:user_id] = nil
```
Comment (CheezItMan) [inline app/controllers/users_controller.rb:32]: Flash notice?

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene |  Good number of commits and good commit messages
Comprehension questions |  Check, I'm happy to show you this week how to troubleshoot Heroku deployments a bit.  It usually varies quite a bit, but I suggest practicing using Heroku logs.  
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  Check
Views are well-organized (DRY, use of semantic HTML, use of partials) |  Check
Errors are reported to the user |  Check
Business logic lives in the models |  Check
Models are thoroughly tested, including relations, validations and any custom logic |  See my notes, but nice work
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work |  See my notes on testing
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional |  Check
**Wave 2 - Users and Votes** |
Users can log in and log out |  Check
The ID of the current user is stored in the session |  Check
A user cannot vote for the same media more than once |  Check
All media lists are ordered by vote count |  Check
Splash page contains a media spotlight |  Check
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users |  Check
Individual user pages and the user list are present |  Check
**Optional - Styling** |
Bootstrap is used appropriately |  Check
Look and feel is similar to the original |  Check, nice work imitating the original
**Overall** |  Nice work, you hit the learning goals for the project.  See my notes, particularly on testing.  You need to be a bit more through when writing your tests, particularly on Controller tests.  Let me know if you have questions.
